### DISCUSSION TITLE
Redirect xonsh lib subprocess call

### DISCUSSION THREAD

**Discussion by eugenesvk** [xonsh/xonsh#5525] [Q&A] [answered]
In my aliases of various cli tools I'm using xonsh lib's subprocess calls (let's call them `xsubproc` to differentiate from python's native `subprocess`) and would like stdout redirection to work so that `myalis > outfile.txt` works
So simply executing as the last command in the alias function
`subprocess.run(my_cmd, stdout=sys.stdout).stdout` seems to work, but then using the lib's 
`xsubproc.run(my_cmd, stdout=sys.stdout).stdout` fails as it doesn't support the extra parameters

Is this a similar issue to https://github.com/xonsh/xonsh/issues/5512 or maybe related to a much broader https://github.com/xonsh/xonsh/issues/5444 (or maybe just the same as this stderr discussion https://github.com/xonsh/xonsh/discussions/5225)?

Or is this a much simpler separate issue which should be opened as such?
Or is there an existing solution to this?

Example alias
```py
import subprocess
from builtins import __xonsh__  # XonshSession (${...} is '__xonsh__.env')
from xonsh.lib        import subprocess as xsubproc
import sys

@aliases.register('a')
def _a():
  cmd = __xonsh__.commands_cache.lazy_locate_binary('example_cli_with_list.exe', ignore_alias=True)
  # subprocess.run([cmd] + ['list'],stdout=sys.stdout).stdout # works
  xsubproc  .run([cmd] + ['list'],stdout=sys.stdout).stdout # fails, unexpected keyword arg
# a > o.txt
```

**Comment by anki-code** [comment]
I think it's more about this - https://github.com/xonsh/xonsh/discussions/5316#discussioncomment-9803773. And may be https://github.com/xonsh/xonsh/issues/5512#issuecomment-2182140516. More trace is welcome!

**Reply by eugenesvk** [reply]
> Back to your question I prefer to say that using xonsh API is recommended. It was covered by tests. When something going wrong you can create an issue around API call.

Would love to do that if only it had that extra argument...

**Reply by anki-code** [reply]
Which?

**Comment by eugenesvk** [comment]
The `stdout=sys.stdout` (and I guess, `stderr` as well)

`xonsh.lib.subprocess.run(cmd, cwd=None, check=False)`

**Reply by anki-code** [reply]
As I understand you're trying to create workaround for #5512. But why not just fix it?

**Reply by eugenesvk** [reply]
Sorry, I don't understand the inner workings of these subprocess calls to answer, hence posted this discussion. Maybe #5512 is exactly the same issue even though it doesn't call `xonsh.lib.subprocess.run` directly, so if it's fixed it would also resolve my issue without introducing any extra explicit parameters

**Reply by anki-code** [reply, marked as answer]
You want to redirect the output of callable alias into a file. Xonsh already has machinery for this and it works in case of pipe, see examples in https://github.com/xonsh/xonsh/issues/5512#issuecomment-2176157599. So I classify uncaptured subprocess command output as a bug and we need to fix #5512.

PS: `xonsh.lib.subprocess.run` calls `subproc_captured_hiddenobject` (i.e. `![]`) inside.

**Reply by eugenesvk** [reply]
After digging more I think this bug is indeed what lead me astray, and despite #5512 being closed that part is still unfixed, right? (but I'll test using the other subproc function)

```py
exe = R"C:\path\to\exe\with\list\command.exe"
cmd = [exe] + ['list']

import time
import datetime
from xonsh.built_ins import subproc_captured_object, subproc_captured_hiddenobject
@aliases.register('sco')
def _sco():
  print(f"→→→ sleep {datetime.datetime.now()}  subproc_captured_object")
  time.sleep(1)
  cmdp = subproc_captured_object(cmd)
  print(f"←←← cmdp = {cmdp}")
  print(f"←←← sleep {datetime.datetime.now()}")

@aliases.register('scoh')
def _scoh():
  print(f"→→→ sleep {datetime.datetime.now()}  subproc_captured_hiddenobject")
  time.sleep(1)
  cmdp = subproc_captured_hiddenobject(cmd)
  print(f"←←← cmdp = {cmdp}")
  print(f"←←← sleep {datetime.datetime.now()}")
```

here `sco > sco` works while `scoh > scoh` fails to redirect

**Reply by anki-code** [reply]
Both as expected and `subproc_captured_hiddenobject` is `$[]` - uncaptured hidden object. Please read https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md#operators

**Comment by anki-code** [comment]
@eugenesvk I prepared the fix - https://github.com/xonsh/xonsh/pull/5527

**Reply by eugenesvk** [reply]
Ok, so I've tried the latest 18 version and the second one still outputs to the terminal, not file with 
`print(xsubproc  .run([cmd] + ['list']).stdout)`
`xsubproc  .run([cmd] + ['list']).stdout`
(tried `$XONSH_CAPTURE_ALWAYS=True`)

Maybe a more general question:
If I have a python script file
```py
import time
import datetime
print(f"pre sleep {datetime.datetime.now()}")
time.sleep(1)
print(f"pos sleep {datetime.datetime.now()}")
```

how can I make an alias `myals` to run this file that would
- when run `myals` print commands in real time (guess `@unthreadable` deals with this, it runs in the foregrond)
- be redirectable `myals > out.txt`

**Reply by anki-code** [reply]
Please make more clear example e.g. 
```
echo 1
# 1
do-thing
# Expected: 2
# etc
```

**Reply by eugenesvk** [reply]
- create a `test.py` file with
```py
#!/usr/bin/env python.exe
import time
import datetime
print(f"pre sleep {datetime.datetime.now()}")
time.sleep(1)
print(f"pos sleep {datetime.datetime.now()}")
```

Create an alias `myals` that would run this file so that 
- `myals`
  - prints `pre sleep` right away
  - waits
  - prints `pos sleep`
   - to this file in xonsh that would
- `myals > out.txt` prints nothing, but saves `pre sleep` and `pos sleep` to file (with timings 1 second apart)

**Reply by anki-code** [reply]
This works to me:

```xsh
cd /tmp
echo @("""
import time
import datetime
print(f"pre sleep {datetime.datetime.now()}")
time.sleep(1)
print(f"pos sleep {datetime.datetime.now()}")
""") > run.py

@aliases.register
def _run():
    python /tmp/run.py
    
run > run.log
cat run.log
# pre sleep 2024-07-19 00:20:52.818751
# pos sleep 2024-07-19 00:20:53.824266
```
If you want to have both output and streaming output to file it's a new feature I think.

**Reply by eugenesvk** [reply]
This doesn't work, `a > a` doesn't redirect output to a file (if you adjust `@aliases.register('a')`

This basic bash does it as expected. Save to `a.sh` and alias as `alias aa = ~/.local/bin/a.sh`
```sh
#!/bin/bash
echo "pre" $(date "+%S")
sleep 1
echo "pre" $(date "+%S")
```

Is there some fundamental difference in xonsh that makes these things not work as expected?

**Reply by anki-code** [reply]
This is working as expected to me in all cases:
```xsh
xonsh -V
# xonsh/0.18.2
xonsh --no-env
```
Create script:
```xsh
cd /tmp
echo @("""
#!/bin/bash
echo "pre" $(date "+%S")
sleep 1
echo "pre" $(date "+%S")
""") > a.sh
chmod +x a.sh
```
In case with string:
```xsh
aliases['a'] = '/tmp/a.sh'

a > a.log
cat a.log
# pre 48
# pre 49
```
In case with callable alias:
```xsh
@aliases.register('a')
def _a():
    /tmp/a.sh

a > a.log
cat a.log
# pre 16
# pre 17
```
In case with return_command alias:

```xsh
@aliases.register('a')
@aliases.return_command
def _a(args):
    return ['/tmp/a.sh'] + args

a > a.log
cat a.log
# pre 3
# pre 4
```
